### PR TITLE
fix(deep-semgrep): Propagate taint through method calls of objects

### PR DESCRIPTION
**Pull request by emjin**
For deep-semgrep

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: semgrep-core/src/tainting/Dataflow_tainting.ml [hunk 1]
```diff
@@ -479,14 +479,15 @@ let check_function_signature env fun_exp args_taints =
...
+                       function. *)
                     id_info =
                       {
-                        G.id_resolved =
-                          {
-                            contents =
-                              Some ((G.ImportedEntity _ | G.ResolvedName _), _);
-                          };
-                        _;
+                        G.id_resolved = { contents = Some _; }; _;
```
Comment (IagoAbal) [inline semgrep-core/src/tainting/Dataflow_tainting.ml]: Now I wished somebody had asked me _why_ I was introducing this restriction at the time I did it. :D

**Review by IagoAbal — APPROVED** [review]

File: semgrep-core/src/tainting/Dataflow_tainting.ml (hunk 1 above)
Comment (IagoAbal) [reply semgrep-core/src/tainting/Dataflow_tainting.ml]: I actually think we should probably not even case-split on `fun_exp` at all. If the function expression is not being resolved (and trying to resolve it's probably cheap?) then we just return `None` and that makes Semgrep take the fall-back assumption (assume it's tainted). If you want you can just do that, and drop the `TODO` note.

Comment (emjin) [reply semgrep-core/src/tainting/Dataflow_tainting.ml]: I found a reason this is bad so I'm motivated to make that fix now!

File: semgrep-core/src/tainting/Dataflow_tainting.ml [hunk 2]
```diff
@@ -471,30 +471,17 @@ let check_function_signature env fun_exp args_taints =
...
-              _;
-            };
+        e = Fetch { base = Var { ident; _ }; rev_offset = []; _ };
         eorig = SameAs eorig;
         _;
+      } )
+  | ( Some hook,
+      (* Case `$X. ... .$F()` *)
+      {
+        e = Fetch { base = _; rev_offset = Dot { ident; _ } :: _; _ };
```
Comment (IagoAbal) [inline semgrep-core/src/tainting/Dataflow_tainting.ml]: Why do you need to distinguish those two cases?

Comment (emjin) [reply semgrep-core/src/tainting/Dataflow_tainting.ml]: It's for `ident`, so that we can have the token of the function. I can clarify

Comment (emjin) [reply semgrep-core/src/tainting/Dataflow_tainting.ml]: Actually, I'll just perform that split lower since we only need it in one case

**Pull request merged**